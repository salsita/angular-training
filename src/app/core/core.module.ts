import { HttpClientModule } from '@angular/common/http';
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { ApiModule } from '@salsita/ng-api';
import { CrudModule } from '@salsita/ng-crud';
import { EntityRepositoryModule } from '@salsita/ng-entities';
import { RouterModule as RouterStoreModule } from '@salsita/ng-router';

import { environment } from '../../environments/environment';
import { SharedModule } from '../shared/shared.module';
import { ApiErrorToastComponent } from './api/api-error-toast/api-error-toast.component';
import { ApiLoaderComponent } from './api/api-loader/api-loader.component';

@NgModule({
  imports: [
    HttpClientModule,

    ApiModule.forRoot({ storeNamespace: 'api', baseUrl: environment.apiBaseUrl }),
    CrudModule.forRoot({ storeNamespace: 'crud' }),
    EntityRepositoryModule.forRoot({ storeNamespace: 'entityRepository' }),
    RouterStoreModule.forRoot({ storeNamespace: 'router' }),
    SharedModule
  ],
  declarations: [ApiErrorToastComponent, ApiLoaderComponent],
  exports: [ApiErrorToastComponent, ApiLoaderComponent, ApiModule, HttpClientModule]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        `CoreModule has already been loaded. Import Core modules in the AppModule only.`
      );
    }
  }
}
