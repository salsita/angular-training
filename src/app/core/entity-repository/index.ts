export {
  EntityRepositoryModuleConfig,
  EntityRepositoryState,
  EntityRepositoryStateRoot,
  SingleEntityRepository
} from './entity-repository.interfaces';
export { EntityRepositorySelectors } from './entity-repository.selectors';
export { EntityRepositoryService } from './entity-repository.service';
export { EntityRepositoryModule } from './entity-repository.module';
