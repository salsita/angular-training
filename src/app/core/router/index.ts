export {
  FlatRouterStateSnapshot,
  RouterModuleConfig,
  RouterState,
  RouterStateRoot
} from './router.interfaces';
export { RouterSelectors } from './router.selectors';
export { CustomRouterStateSerializer } from './router.serializer';
export { RouterModule } from './router.module';
