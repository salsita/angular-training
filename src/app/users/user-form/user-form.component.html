<form [formGroup]="form" (ngSubmit)="onSubmit()" (submit)="$event.stopPropagation()">
  <label>First Name</label>
  <input
    formControlName="firstName"
    type="text"
    placeholder="John"
  />
  <label>Last Name</label>
  <input
    formControlName="lastName"
    type="text"
    placeholder="Doe"
  />
  <label>E-mail</label>
  <input
    formControlName="email"
    type="email"
    placeholder="email@email.com"
  />

  <h2>Skills</h2>
  <ul>
    <li formArrayName="skills" *ngFor="let skillControl of form.get('skills').controls; let i = index">
      <select [formControlName]="i">
        <option *ngFor="let skill of skills" [ngValue]="skill">{{skill.skill}}</option>
      </select>
      <button (click)="removeSkill(i)">x</button>
    </li>
  </ul>
  <button type="button" (click)="addSkill()">Add skill</button>
  <br>
  <button type="submit">Save</button>
</form>
