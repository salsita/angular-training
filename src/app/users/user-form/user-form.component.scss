app-form-field {
  display: block;
}
