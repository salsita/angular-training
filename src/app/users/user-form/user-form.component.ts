import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormArray, FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
import { Skill, User } from '../users.interfaces';

@Component({
  selector: 'app-user-form',
  templateUrl: './user-form.component.html',
  styleUrls: ['./user-form.component.scss']
})
export class UserFormComponent implements OnInit {
  @Input() skills: Skill[] = [];
  @Input()
  user: User = {
    id: '',
    firstName: '',
    lastName: '',
    email: '',
    skills: []
  };

  @Output() submit = new EventEmitter<User>();

  form: FormGroup;

  constructor(private formBuilder: FormBuilder) {}

  addSkill() {
    (this.form.get('skills') as FormArray).push(new FormControl());
  }

  ngOnInit() {
    this.form = this.formBuilder.group({
      firstName: [this.user.firstName, Validators.required],
      lastName: [this.user.lastName, Validators.required],
      email: [this.user.email, Validators.required],
      skills: this.formBuilder.array(this.user.skills)
    });
  }

  removeSkill(index: number) {
    (this.form.get('skills') as FormArray).removeAt(index);
  }

  onSubmit() {
    this.submit.emit(this.form.value);
  }
}
