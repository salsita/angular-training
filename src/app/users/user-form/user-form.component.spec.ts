import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';

import { UserFormComponent } from './user-form.component';

describe('UserFormComponent', () => {
  let component: UserFormComponent;
  let fixture: ComponentFixture<UserFormComponent>;

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        imports: [ReactiveFormsModule],
        declarations: [UserFormComponent],
        schemas: [NO_ERRORS_SCHEMA]
      }).compileComponents();
    })
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(UserFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
