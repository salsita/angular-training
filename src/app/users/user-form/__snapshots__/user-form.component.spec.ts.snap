// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFormComponent should add skill 1`] = `
<app-user-form
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  skills={[Function Array]}
  submit={[Function EventEmitter]}
  user={[Function Object]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    
    
    <app-form-field
      label="First Name"
    />
    
    
    <app-form-field
      label="Last Name"
    />
    
    
    <app-form-field
      label="E-mail"
    />
    
  
    
    <h2>
      Skills
    </h2>
    
    
    
    <ul>
      
      
      
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="0"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="1"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="2"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      
    
    </ul>
    
    
    <button
      type="button"
    >
      Add skill
    </button>
    
    
    <br />
    
    
    <button
      type="submit"
    >
      Save
    </button>
    
  
  </form>
  
</app-user-form>
`;

exports[`UserFormComponent should remove skill 1`] = `
<app-user-form
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  skills={[Function Array]}
  submit={[Function EventEmitter]}
  user={[Function Object]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    
    
    <app-form-field
      label="First Name"
    />
    
    
    <app-form-field
      label="Last Name"
    />
    
    
    <app-form-field
      label="E-mail"
    />
    
  
    
    <h2>
      Skills
    </h2>
    
    
    
    <ul>
      
      
      
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="0"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      
    
    </ul>
    
    
    <button
      type="button"
    >
      Add skill
    </button>
    
    
    <br />
    
    
    <button
      type="submit"
    >
      Save
    </button>
    
  
  </form>
  
</app-user-form>
`;

exports[`UserFormComponent should render user form 1`] = `
<app-user-form
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  skills={[Function Array]}
  submit={[Function EventEmitter]}
  user={[Function Object]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    
    
    <app-form-field
      label="First Name"
    />
    
    
    <app-form-field
      label="Last Name"
    />
    
    
    <app-form-field
      label="E-mail"
    />
    
  
    
    <h2>
      Skills
    </h2>
    
    
    
    <ul>
      
      
      
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="0"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      <li
        class="ng-untouched ng-pristine ng-valid"
        formarrayname="skills"
        ng-reflect-name="skills"
      >
        
        
        <select
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-name="1"
        >
          
          
          
          <option
            ng-reflect-ng-value="[object Object]"
            value="0: Object"
          >
            JavaScript
          </option>
          <option
            ng-reflect-ng-value="[object Object]"
            value="1: Object"
          >
            PostgreSQL
          </option>
          
        
        </select>
        
        
        <button>
          x
        </button>
        
      
      </li>
      
    
    </ul>
    
    
    <button
      type="button"
    >
      Add skill
    </button>
    
    
    <br />
    
    
    <button
      type="submit"
    >
      Save
    </button>
    
  
  </form>
  
</app-user-form>
`;
